python/ctsm/modify_fsurdat/modify_fsurdat.py

```diff
@@ -0,0 +1,288 @@
+        If user has specified lai, sai, hgt_top, hgt_bot, replace these with
+        values selected by the user for dom_nat_pft
+
+        Arguments
+        ---------
+        dom_nat_pft:
+            (int) User's entry of PFT to be set to 100% everywhere
+        lai:
+            (float) User's entry of MONTHLY_LAI for their dom_nat_pft
+        sai:
+            (float) User's entry of MONTHLY_SAI for their dom_nat_pft
+        hgt_top:
+            (float) User's entry of MONTHLY_HEIGHT_TOP for their dom_nat_pft
+        hgt_bot:
+            (float) User's entry of MONTHLY_HEIGHT_BOT for their dom_nat_pft
+        """
+
+        # initialize 3D variable
+        self.file['PCT_NAT_PFT'][:,:,:] = self.setvar(
+            self.file['PCT_NAT_PFT'][:,:,:], 0)
+        # set 3D variable value for dom_nat_pft
+        self.file['PCT_NAT_PFT'][dom_nat_pft,:,:] = self.setvar(
+            self.file['PCT_NAT_PFT'][dom_nat_pft,:,:], 100)
+
+        # dictionary of 4d variables to loop over
+        vars_4d = {'MONTHLY_LAI': lai,
+                   'MONTHLY_SAI': sai,
+                   'MONTHLY_HEIGHT_TOP': hgt_top,
+                   'MONTHLY_HEIGHT_BOT': hgt_bot}
+        for var_name, var in vars_4d.items():
+            if var is not None:
+                self.set_lai_sai_hgts(dom_nat_pft=dom_nat_pft,
+                                      var_name=var_name, var=var)
+
+
+    def set_lai_sai_hgts(self, dom_nat_pft, var_name, var):
+        """
+        Description
+        -----------
+        If user has specified lai, sai, hgt_top, hgt_bot, replace these with
+        values selected by the user for dom_nat_pft. Else do nothing.
+        """
+        if dom_nat_pft == 0:  # bare soil: var must equal 0
+            var = [0] * 12
+        if len(var) != 12:
+            errmsg = 'Error: Variable should have exactly 12 ' \
+                     'entries in the configure file: ' + var_name
+            abort(errmsg)
+        for mon in self.file.time - 1:  # loop over 12 months
+            # set 4D variable to value for dom_nat_pft
+            self.file[var_name][mon,dom_nat_pft,:,:] = self.setvar(
+                self.file[var_name][mon,dom_nat_pft,:,:], var[int(mon)])
+
+
+    def zero_nonveg(self):
+        """
+        Description
+        -----------
+        Set all landunit weights to 0 except the natural vegetation landunit.
+        Set that one to 100%.
+
+        Arguments
+        ---------
+        """
+
+        self.file['PCT_NATVEG'][:,:] = 100
+        self.file['PCT_CROP'][:,:] = 0
+        self.file['PCT_LAKE'][:,:] = 0
+        self.file['PCT_WETLAND'][:,:] = 0
+        self.file['PCT_URBAN'][:,:] = 0
+        self.file['PCT_GLACIER'][:,:] = 0
+
+
+    def setvar(self, var, val):
+        """
+        Set variable var to value val in the *rectangle* domain, which
+        is denoted as *other* in the *not_rectangle* domain.
+        """
+
+        var = var.where(self.not_rectangle, other=val)
+
+        return var
```

Comment:
- [ ] As mentioned above, there would be better [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)) in this class if this function did the setting of the file variable – changing its interface to `def setvar(self, varname, val)` and then actually doing the `self.file[varname] = ...` itself. Sorry that I didn't explain my thinking clearly before.

``` python
def setvar(self, varname, val):
"""
Sets the value for a single-level field
"""
    self.file[varname] = self.file[varname].where(self.not_rectangle, other=val)
```

- [ ] Also, if I'm right in my comment above that this doesn't work right for setting a 3-d variable, then you could add a check in this function that `self.file[varname]` has the same size/shape as `self.not_rectangle`

- [ ] I see that you may need to then have a separate version of this function for multi-level variables, like:

``` python
def setvar_onelevel(self, varname, val, level):
"""
Sets one level of a multi-level field
"""
    self.file[varname][level,:,:] = self.file[varname][level,:,:].where(self.not_rectangle, other=val)
```

- [ ] And similarly for a version that sets one slice of a 4-d field (as needed by `set_lai_sai_heights`); I'm not sure off-hand what the best naming would be here

One other advantage of redefining the function in this way is: We'll need this to work on surface datasets that are on unstructured as well as structured grids. The former will imply that there is only one spatial dimension, not two. You should then be able to make changes in just a small number of places to handle files with either one or two spatial dimension(s).

Reply:
I think I have addressed the comments in this post now.

I did not check whether my code the way I had it would work for a 3d variable. I just moved all 3d and 4d vars into for loops.